theano/tensor/slinalg.py

```diff
@@ -202,10 +202,16 @@ def make_node(self, A, b):
         b = as_tensor_variable(b)
         assert A.ndim == 2
         assert b.ndim in [1, 2]
-        otype = tensor.tensor(
+        if ((A.dtype == 'float32' and b.dtype == 'float32')
+                or (A.dtype in ['int8', 'int16'] and b.dtype == 'float32')
+                or (b.dtype in ['int8', 'int16'] and A.dtype == 'float32')):
+            o_dtype = 'float32'
+        else:
+            o_dtype = 'float64'
+        x = tensor.tensor(
             broadcastable=b.broadcastable,
-            dtype=(A * b).dtype)
```

Comment:
Yes it does, e.g. `matrix('int32') * matrix('int32') => matrix('int32')` but solve will give a float. A more compact implementation could be to actually perform scipy.linalg.solve on very small matrices (e.g. 1x1) with correct dtypes and get the dtype of the output. This would probably slow down building the graph a little bit compared to conditionals.